File: Java/src/SerializeAndDeserializeBinaryTree.java
import DataStructure.TreeNode;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return "X.";
        }
        return root.val + "." + serialize(root.left) + serialize(root.right);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] nodeVals = data.split("\\.");
        Queue<String> nodeValList = new LinkedList<>(Arrays.asList(nodeVals));
        return process(nodeValList);
    }

    private TreeNode process(Queue<String> list) {
        String nodeVal = list.poll();
        if (nodeVal == null || nodeVal.equals("X")) {
            return null;
        }
        TreeNode node = new TreeNode(Integer.parseInt(nodeVal));
        node.left = process(list);
        node.right = process(list);
        return node;
    }

}

public class SerializeAndDeserializeBinaryTree {
    public static void main(String[] args) {
        TreeNode node1 = new TreeNode(1);
        TreeNode node2 = new TreeNode(2);
        TreeNode node3 = new TreeNode(3);
        TreeNode node4 = new TreeNode(4);
        TreeNode node5 = new TreeNode(5);
        node1.left = node2;
        node1.right = node3;
        node2.left = node4;
        node2.right = node5;
        Codec codec = new Codec();
        System.out.println(codec.serialize(node1));
    }
}

File: Python/Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    delimiter = "_"
    none_str = "None"

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """

        def helper(node: TreeNode, s: str):
            if not node:
                s += self.none_str + self.delimiter
            else:
                s += str(node.val) + self.delimiter
                s = helper(node.left, s)
                s = helper(node.right, s)
            return s

        return helper(root, "")

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """

        def helper(strs):
            if strs[0] == self.none_str:
                strs.pop(0)
                return None
            node = TreeNode(int(strs[0]))
            strs.pop(0)
            node.left = helper(strs)
            node.right = helper(strs)
            return node

        strs = data.split(self.delimiter)
        return helper(strs)

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
